### ISSUE TITLE
Is it possible to use leader-line with other draggable library?

### ISSUE THREAD

**Issue by jorgemonte** [anseki/leader-line#38]
Hi again Anseki!

One more time, thanks and congratulations for creating and sharing with us this amazing tool. I am sorry to bother you again with doubts and questions.
I was wondering if would be possible to use the lines with another draggable library. I found one called Displace.js that is very simple to use and lightweight, and it was easier to get for example my draggable element to go outside its div. I was trying to do the same with PlainDraggable and didn't get, unfortunately. 

I need to build something like this: 
https://codepen.io/A8-XPs/pen/JwWLOd

They need to be interactive for the user, which will be able to hold a circle and drag wherever he wants on the website, and the lines need to follow the circle, creating a deformation in the shape (this is intended). I can't find a way to make the circles to go outside their parent div (this is the first difficult). Also, the circles must be still draggable after we bring it outside of its parent div.

I apologize again to disturb you and would like to kindly request the possibility of help with samples.
I have those shapes in SVG but also considered to make them with span elements in absolute position, I planned to make the lines (paths) to be created with the leader-line and of course the visual result would be similar but not exactly the same (I can define the svg paths to be transparent).

I had difficult to create shared draggable elements, as the example that I sent you some of the circles are the start point for multiple items. Or, in the oval shape example the start point of a line is the end point to another. 

Thank you very much!

**Comment by jorgemonte** [comment]
Hello @anseki 
Thank you again, there's no problem, my english is also average hehe.
Yes, I need the element to be able to go outsite its canvas or div, I think the solution you suggested will work perfectly! The other difficult that I have is, for example: I need a new line going from the red circle to a new circle (let's say, a green one), and then another line starting at the green circle and returning to the blue one (so, the blue circle is the start point of the first line, but at the same time it is the last point of the cycle). I need to create shapes similar to the ones that I sent the link. Is it possible to make?

I got to create 3 elements and having 2 lines, but when I moved the second circle, it wasn't draggable anymore. Thank you again!

**Comment by anseki** [comment]
Is this it?
https://jsfiddle.net/zb7Lp5o9/2/

**Comment by jorgemonte** [comment]
Yes my friend! Exactly that hehe. You are a javascript ninja, I am greatful for your help.

I have only 1 more thing to ask you if is possible to do. When the line passes from the side to the top (for example), it goes very sudden. Is there a way to make the line pass around the circles more smoothly? Instead of having only the 4 points (top, left, bottom, right), would be possible to the line move around the circles more smooth when we change their position? Thank you again !

**Comment by anseki** [comment]
You can use [`pointAnchor`](https://anseki.github.io/leader-line/#pointanchor) attachment to adjust the point that the line is connected.
For example:
https://jsfiddle.net/xj54b71d/

**Comment by jorgemonte** [comment]
This is amazing, it's incredible. It's exactly what I meant. So, having this smooth movement included at the other specs that you showed before it's possible? (having multiple connected elements) or, it is just possible to have this result linking 2 elements?
I am really impressed with the power of this library, one more time: congratulations my friend.

**Comment by anseki** [comment]
The example I wrote might look complex, but its logic is easy.
You can get an angle, from coordinates of two elements.
Move from the center of each circle with the angle and the radius of that circle, then, that is the point.

**Comment by jorgemonte** [comment]
Yes but, for example, if I have 3 circles connected with lines (as you did in the previous example), if I move 1 circle, it will affect the position of 2 lines instead of one. This is the part I still can't figure out how it will be coded to create the smooth changing of the line position.

**Comment by anseki** [comment]
You have to just only update another line in like manner, in the `onMove` function.
All lines have two elements. You can get the points via those two elements of each line.

**Comment by jorgemonte** [comment]
I will try to build it following the structure that you helped to make. Thank you! Wish you great success!

**Comment by jorgemonte** [comment]
Hello @anseki  how are you?
I am sorry to disturb again, lol. I am trying to apply both of the things that we talked last time (The multiple elements linked and the smooth movement of the line around the circles) and I don't know the reason why it doesnt work, it requires a function that I don't get.
Could you help me to take a look what is the point that I am missing?
Here is the code that I did so far:
https://jsfiddle.net/jorgemonte/9as80pck/1/
Thank you!

**Comment by anseki** [comment]
Hi @jorgemonte, I hope I can be of assistance. :smile:

I opened your JSFiddle page, but that did not work. It seems that it could not load libraries. That is, you mistook URLs that are specified for "Resources" fields.

The most basic and general way to solve a problem in web browser is inspection by "Developer Tools" of web browser. You can open that by pressing `F12` key.
In "Console" tab in the Developer Tools when the JSFiddle page was opened, you will see error messages. Those say that the `leader-line.min.js` and `plain-draggable.min.js` could not be loaded.

Maybe you specified the `plain-draggable.min.js` and `leader-line.min.js` as URL (i.e. not only file name).
By HTML rule, a URL that has not path is regarded as a path to current directory. That is, the URL path to a directory in jsfiddle.net server. Of course those files don't exist in jsfiddle.net server, then those errors are thrown.

Therefore, you have to specify correct URLs.
You can use CDN service that can be accessed from anywhere. For example, https://www.jsdelivr.com/ is famous.
Input "leader-line" in the "search all of npm" field, and you will get a URL of CDN File:
https://cdn.jsdelivr.net/npm/leader-line@1.0.5/leader-line.min.js
You can specify those URLs in "Resources" fields in the JSFiddle page.

**Comment by jorgemonte** [comment]
Hi anseki!

It is weird, it worked yesterday on my jsfiddle when I created it in the other computer hehe, I used the CDN url. maybe I got the wrong url hehe. I will copy the code here maybe it is easier to share?

I am sorry to bother you again 

HTML:
```
<svg id="view">
  <circle id="blue" r="16" cx="64" cy="64"/>
  <circle id="red" r="16" cx="160" cy="160"/>
  <circle id="green" r="16" cx="64" cy="256"/>
</svg>
```

JS:

```
var OFFSET_X = 16,
  OFFSET_Y = 16,
  ANCHOR_RADIUS = 16,

  element0 = document.getElementById('blue'),
  element1 = document.getElementById('red'),
  element2 = document.getElementById('green'),
  line = new LeaderLine(element0, element1,
    {path: 'straight', startPlug: 'disc', endPlug: 'disc'}),
  draggable0, draggable1;
  

function update() {
  function offsetPoint(point, angle, len, basePoint) {
    return {
      x: point.x + Math.cos(angle) * len - basePoint.x,
      y: point.y + Math.sin(angle) * len * -1 - basePoint.y
    };
  }

  var xy0 = {x: draggable0.left, y: draggable0.top},
    xy1 = {x: draggable1.left, y: draggable1.top},
    point0 = {x: xy0.x + OFFSET_X, y: xy0.y + OFFSET_Y},
    point1 = {x: xy1.x + OFFSET_X, y: xy1.y + OFFSET_Y},
    angle = Math.atan2(point0.y - point1.y, point1.x - point0.x),
    offsetPoint0 = offsetPoint(point0, angle, ANCHOR_RADIUS, xy0),
    offsetPoint1 = offsetPoint(point1, angle + Math.PI, ANCHOR_RADIUS, xy1);

  line.setOptions({
    start: LeaderLine.pointAnchor(element0, offsetPoint0),
    end: LeaderLine.pointAnchor(element1, offsetPoint1)
  });
}

lineTwo = new LeaderLine(element1, element2,
    {path: 'straight', startPlug: 'disc', endPlug: 'disc'}),
  draggable1, draggable2;
function update() {
  function offsetPoint(point, angle, len, basePoint) {
    return {
      x: point.x + Math.cos(angle) * len - basePoint.x,
      y: point.y + Math.sin(angle) * len * -1 - basePoint.y
    };
  }

  var xy0 = {x: draggable1.left, y: draggable1.top},
    xy1 = {x: draggable2.left, y: draggable2.top},
    point0 = {x: xy0.x + OFFSET_X, y: xy0.y + OFFSET_Y},
    point1 = {x: xy1.x + OFFSET_X, y: xy1.y + OFFSET_Y},
    angle = Math.atan2(point0.y - point1.y, point1.x - point0.x),
    offsetPoint0 = offsetPoint(point0, angle, ANCHOR_RADIUS, xy0),
    offsetPoint1 = offsetPoint(point1, angle + Math.PI, ANCHOR_RADIUS, xy1);

  line.setOptions({
    start: LeaderLine.pointAnchor(element1, offsetPoint0),
    end: LeaderLine.pointAnchor(element2, offsetPoint1)
  });
}



draggable0 = new PlainDraggable(element0, {onMove: update});
draggable1 = new PlainDraggable(element1, {onMove: update});
draggable2 = new PlainDraggable(element2, {onMove: update});
update();
```

**Comment by anseki** [comment]
No, the copying here is not easy to share that because it never work at here (i.e. GitHub issue system is not a testing system). Also, I can read your code in the JSFiddle page very easily (just only one click, then the code is shown in beauty viewer), then you have no cause for copying that.
You have to only input correct URLs.

**Comment by jorgemonte** [comment]
Yes, I am sorry, going to do that. Anyways, it is strange that it was working when I saved, I put the correct URL's and it worked well, after that it didn't work anymore. I will try to figure out what is wrong. Thank you!

**Comment by jorgemonte** [comment]
Hi Anseki!

I think that now the link is the right one:
https://jsfiddle.net/jorgemonte/xj54b71d/3/ 

I just put 2 lines, in order to test the experience, the circles are no longer draggable (even if the library is correctly installed, it worked before I added the second line).

**Comment by anseki** [comment]
The most basic and general way to solve a problem in web browser is inspection by "Developer Tools" of web browser. You can open that by pressing F12 key.
In "Console" tab in the Developer Tools when the JSFiddle page was opened, you will see error messages.
See: https://github.com/anseki/leader-line/issues/38#issuecomment-451336383
That says "draggable2 is not defined".

That is, you defined (i.e. `var` statement) `draggable0` and `draggable1`, but you did not define the `draggable2`. The error occurs when you use the `draggable2`.

I say again, you should use the Developer Tools of web browser that is very useful. In other words, all developers need the Developer Tools.
